src/ipc/ipc4/handler.c

```diff
@@ -715,10 +715,6 @@ static int ipc4_process_chain_dma(struct ipc4_message_request *ipc4)
 		return IPC4_SUCCESS;
 	}
 
-	cdma_comp = ipc_get_comp_by_id(ipc, comp_id);
```

Comment:
compare the if part, there are below difference:

1. comp_free(cdma_comp->cd);, this added in line 711, however, missed in following error handling.
2. can we confirm that ipc4_chain_dma_state need twice calling? if line 693 condition meet, if part will call this.
    and following line 722 will also call this, seems a little not so reasonable.

Reply:
@lyakh is this resolved/valid ?

Reply:
@lgirdwood sorry, I don't understand whether that's a comment or a question or a change request? @btian1 can you clarify?

Reply:
The only thing I can state here, is that I'm pretty sure that that code is unneeded, a repeated call to `ipc_get_comp_by_id()` there is unneeded and the repeated check is redundant

Reply:
it is further ask for this function, simple for this patch, it is ok go for merge, I mean we can study and discuss here to add more patches if aligned, or you simply want to go with this single patch?
@RanderWang @ujfalusi please share your comments on below ask:

1. comp_free(cdma_comp->cd);, this added in line 711, however, missed in following error handling.
2. can we confirm that ipc4_chain_dma_state need twice calling? if line 693 condition meet, if part will call this.
and following line 722 will also call this, seems a little not so reasonable.

Reply:
@btian1, I agree that the error handling is wrong but it is not the topic of this PR.
Furthermore: a simple comp_free() is inadequate, it is leaking memory for ChainDMA, see: https://github.com/thesofproject/sof/pull/8767/commits/d2ffc0b227b5213acaa260402b4157833f50c214

We need a new wrapper to properly clean up after ipc4_chain_manager_create() and to be precise after  ipc4_add_comp_dev().